tests/2d_examples/test_2d_shell_collision/src/shell_collision.cpp

```diff
@@ -0,0 +1,277 @@
+	BallMaterial* free_ball_material = new BallMaterial();
+	ElasticSolidParticles 	free_ball_particles(free_ball, free_ball_material);
+
+	FreeBallObserver* free_ball_observer = new FreeBallObserver(sph_system, "FreeBallObserver");
+	BaseParticles 	free_ball_observer_particles(free_ball_observer);
+	//----------------------------------------------------------------------
+	//	Run particle relaxation for body-fitted distribution if chosen.
+	//----------------------------------------------------------------------
+	if (sph_system.run_particle_relaxation_)
+	{
+		//----------------------------------------------------------------------
+		//	Define body relation map used for particle relaxation.
+		//----------------------------------------------------------------------
+		BodyRelationInner* free_ball_inner = new BodyRelationInner(free_ball);
+		//----------------------------------------------------------------------
+		//	Define the methods for particle relaxation.
+		//----------------------------------------------------------------------
+		RandomizePartilePosition  			free_ball_random_particles(free_ball);
+		relax_dynamics::RelaxationStepInner free_ball_relaxation_step_inner(free_ball_inner);
+		//----------------------------------------------------------------------
+		//	Output for particle relaxation.
+		//----------------------------------------------------------------------
+		BodyStatesRecordingToVtu 		write_ball_state(in_output, sph_system.real_bodies_);
+		ReloadParticleIO 		write_particle_reload_files(in_output, { free_ball });
+		//----------------------------------------------------------------------
+		//	Particle relaxation starts here.
+		//----------------------------------------------------------------------
+		free_ball_random_particles.parallel_exec(0.25);
+		write_ball_state.writeToFile(0.0);
+		//----------------------------------------------------------------------
+		//	From here iteration for particle relaxation begines.
+		//----------------------------------------------------------------------
+		int ite = 0;
+		int relax_step = 1000;
+		while (ite < relax_step)
+		{
+			free_ball_relaxation_step_inner.exec();
+			ite += 1;
+			if (ite % 100 == 0)
+			{
+				std::cout << std::fixed << std::setprecision(9) << "Relaxation steps N = " << ite << "\n";
+				write_ball_state.writeToFile(Real(ite) * 1.0e-4);
+			}
+		}
+		std::cout << "The physics relaxation process of ball particles finish !" << std::endl;
+		write_particle_reload_files.writeToFile(0.0);
+		return 0;
+	}
+	//----------------------------------------------------------------------
+	//	Define body relation map.
+	//	The contact map gives the topological connections between the bodies.
+	//	Basically the the range of bodies to build neighbor particle lists.
+	//----------------------------------------------------------------------
+	BodyRelationInner*   free_ball_inner = new BodyRelationInner(free_ball);
```

Comment:
we should try to not use pointers for these cases

Reply:
done